crates/core/src/util.rs

```diff
@@ -262,41 +256,52 @@ pub trait TimeSource {
 ///
 /// The private () field is used to make the struct non-constructible outside of
 /// this module to ensure that the time updater is spawned.
-pub struct CachingSystemTimeSrc(());
+#[derive(Clone, Copy)]
+pub(crate) struct CachingSystemTimeSrc(());
+
+// would be nice to use AtomicU128 and store unix timestamp, unfortunately
+// atomic u128 is not available and that would force us to use SystemTime instead
+static GLOBAL_TIME_STATE: AtomicPtr<Instant> = AtomicPtr::new(std::ptr::null_mut());
 
 impl CachingSystemTimeSrc {
     pub(crate) fn new() -> Self {
+        let mut current_unix_epoch_ts = Instant::now();
+        if GLOBAL_TIME_STATE
+            .compare_exchange(
+                std::ptr::null_mut(),
+                (&mut current_unix_epoch_ts) as *mut _,
+                std::sync::atomic::Ordering::Acquire,
+                std::sync::atomic::Ordering::Relaxed,
+            )
+            .is_ok()
         {
-            let mut time_state = GLOBAL_TIME_STATE.write().unwrap();
-            println!("Checking whether to spawn the time updater");
-            if !time_state.1 {
-                println!("Spawning the time updater");
-                time_state.1 = true; // Set the flag
-                let updater_state = Arc::clone(&GLOBAL_TIME_STATE);
-                spawn(async move {
-                    CachingSystemTimeSrc::update_instant(updater_state).await;
-                });
+            let drop_guard = Arc::new(AtomicBool::new(false));
+            tokio::spawn(Self::update_instant(drop_guard.clone()));
```

Comment:
Won't this spawn a new tokio thread every time `CachingSystemTimeSrc::new()` is called? That was the reason I had a bool flag.

Reply:
That condition is only evaluated once